src/librustc_error_codes/error_codes/E0747.md

```diff
@@ -0,0 +1,4 @@
+Negative impls cannot be default impls. A default impl supplies
+default values for the items within to be used by other impls, whereas
+a negative impl declares that there are no other impls. These don't
+make sense to combine.
```

Comment:
I'm happy to register it as a "design point to be resolved later".

Let me make sure I understand your point, first. I believe what you are proposing is this. Users would be allowed to write

```
default impl !Trait for Foo { }
```

and that would be legal. However, if the `default` impl included any items, that would be illegal?

I guess that the general theory here is a `default impl` with no items *in general* has no real meaning? (Which seems true.) 

I would find this appealing if there were some kind of code-generation or macro examples that were made easier by this by such a rule. I can't really think of any at this time, though.

That said, I don't really agree that a decision here one way or the other has any appreciable impact on the complexity of the language semantics. It seems to me that having various rules about what sorts of modifies can and cannot be combined is a pretty trivial piece of complexity.

The main *benefit* I see of reporting an error is that it steers users in the right direction. As the text above says, I really don't see what a `default impl !Trait` should even mean. *Permitting* it just seems to create the potential for confusion to me. However, a lint would also be a reasonable way to steer users away.

Reply:
Indeed, I think users should be allowed to write `default impl !Trait for Foo { }` because the `impl !Trait for Foo {}` already enforces the type checking condition necessary (that there should be no items in the `impl`). Beyond that, imposing a special rule on `default` is in my view merely a lint, which I would have no issue with. (Although I am skeptical that even a lint would actually be steering users away from a real pitfall, as opposed to one we are imagining could happen. Simply put, doing `default impl !Trait...` seems quite unusual.)

What this would add is a side-condition to the typing rules for `impl`s. No single such side-condition would on its own add notable complexity to the language, but if we have a general design philosophy that encourages these "lint made into error" checks, then the sum total of those side-conditions start weighing down specification.

That said, I'm happy to defer this as an unresolved question, but please do add the FIXME to the code as well.

Reply:
Added to tracking issue.